src/optimizer_interface.jl

```diff
@@ -123,26 +123,45 @@ function set_optimizer(
 end
 
 """
-    optimize!(model::Model;
-              ignore_optimize_hook=(model.optimize_hook === nothing),
-              kwargs...)
+    optimize!(
+        model::Model;
+        ignore_optimize_hook = (model.optimize_hook === nothing),
+        differentiation_backend::MOI.Nonlinear.AbstractAutomaticDifferentiation =
+            MOI.Nonlinear.SparseReverseMode(),
+        kwargs...,
+    )
+
+Optimize the model.
+
+If an optimizer has not been set yet (see [`set_optimizer`](@ref)), a
+[`NoOptimizer`](@ref) error is thrown.
 
-Optimize the model. If an optimizer has not been set yet (see
-[`set_optimizer`](@ref)), a [`NoOptimizer`](@ref) error is thrown.
+If `ignore_optimize_hook == true`, the optimize hook is ignored and the model is
+solved as if the hook was not set. Keyword arguments `kwargs` are passed to the
+`optimize_hook`. An error is thrown if `optimize_hook` is `nothing` and keyword
+arguments are provided.
 
-Keyword arguments `kwargs` are passed to the `optimize_hook`. An error is
-thrown if `optimize_hook` is `nothing` and keyword arguments are provided.
+Pass `differentiation_backend` to set the [`MOI.Nonlinear.AbstractAutomaticDifferentiation`](@ref)
+backend used to compute derivatives of nonlinear programs.
+
+If you require only `:ExprGraph`, it is more efficient to pass
+`differentiation_backend = MOI.Nonlinear.ExprGraphOnly()`.
 """
 function optimize!(
     model::Model;
     ignore_optimize_hook = (model.optimize_hook === nothing),
+    differentiation_backend::MOI.Nonlinear.AbstractAutomaticDifferentiation = MOI.Nonlinear.SparseReverseMode(),
```

Comment:
I'm wondering whether this should be a keyword argument to `optimize!` or something set with a separate function.
For instance, if you call `optimize!` several times then you need to set it every time.
In the future, if the differentiation is done through bridges then this should be passed as an MOI attribute so having a separate function seems preferable in that regards.

Reply:
I don't have a strong preference either, but this seems like something that might change once we know all the use cases. Maybe we should keep this argument as an internal API for now (e.g., name it `_differentiation_backend` and document accordingly) so people have a chance to experiment?

Reply:
Okay, I've renamed to `_differentiation_backend`.